fem/bilininteg_convection.cpp

```diff
@@ -163,292 +147,668 @@ void PAConvectionApply2D(const int NE,
+                  BBGu_ += bx * BGu[dz][qy][qx];
+               }
+               GBBu[qz][qy][qx] = GBBu_;
+               BGBu[qz][qy][qx] = BGBu_;
+               BBGu[qz][qy][qx] = BBGu_;
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      double (*DGu)[max_Q1D][max_Q1D] = (double (*)[max_Q1D][max_Q1D])sm3;
+      MFEM_FOREACH_THREAD(qz,z,Q1D)
+      {
+         MFEM_FOREACH_THREAD(qy,y,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qx,x,Q1D)
+            {
+               const double O1 = op(qx,qy,qz,0,e);
+               const double O2 = op(qx,qy,qz,1,e);
+               const double O3 = op(qx,qy,qz,2,e);
 
-   });
+               const double gradX = BBGu[qz][qy][qx];
+               const double gradY = BGBu[qz][qy][qx];
+               const double gradZ = GBBu[qz][qy][qx];
 
+               DGu[qz][qy][qx] = (O1 * gradX) + (O2 * gradY) + (O3 * gradZ);
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      double (*BDGu)[max_Q1D][max_Q1D] = (double (*)[max_Q1D][max_Q1D])sm4;
+      MFEM_FOREACH_THREAD(qx,x,Q1D)
+      {
+         MFEM_FOREACH_THREAD(qy,y,Q1D)
+         {
+            MFEM_FOREACH_THREAD(dz,z,D1D)
+            {
+               double BDGu_ = 0.0;
+               for (int qz = 0; qz < Q1D; ++qz)
+               {
+                  const double w  = Bt(dz,qz);
+                  BDGu_ += w * DGu[qz][qy][qx];
+               }
+               BDGu[dz][qy][qx] = BDGu_;
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      double (*BBDGu)[max_D1D][max_Q1D] = (double (*)[max_D1D][max_Q1D])sm5;
+      MFEM_FOREACH_THREAD(dz,z,D1D)
+      {
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            MFEM_FOREACH_THREAD(dy,y,D1D)
+            {
+               double BBDGu_ = 0.0;
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  const double w  = Bt(dy,qy);
+                  BBDGu_ += w * BDGu[dz][qy][qx];
+               }
+               BBDGu[dz][dy][qx] = BBDGu_;
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      MFEM_FOREACH_THREAD(dz,z,D1D)
+      {
+         MFEM_FOREACH_THREAD(dy,y,D1D)
+         {
+            MFEM_FOREACH_THREAD(dx,x,D1D)
+            {
+               double BBBDGu = 0.0;
+               for (int qx = 0; qx < Q1D; ++qx)
+               {
+                  const double w  = Bt(dx,qx);
+                  BBBDGu += w * BBDGu[dz][dy][qx];
+               }
+               y(dx,dy,dz,e) = BBBDGu;
+            }
+         }
+      }
+   });
 }
 
+void ConvectionIntegrator::AssemblePA(const FiniteElementSpace &fes)
+{
+   // Assumes tensor-product elements
+   Mesh *mesh = fes.GetMesh();
+   const FiniteElement &el = *fes.GetFE(0);
+   ElementTransformation &Trans = *fes.GetElementTransformation(0);
+   const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el, Trans);
+   const int dims = el.GetDim();
+   const int symmDims = dims; // 1x1: 1, 2x2: 3, 3x3: 6
```

Comment:
Symmdims? I think the comment should be along the lines of ``// 1x1: 1, 2x1: 2, 3x1: 3``.

Reply:
I removed the comment.